README.md

```diff
@@ -25,64 +39,124 @@ example or documentation you're looking at just in case there's a mismatch that
-**ng serve**. This will trigger the various tools in the
+Once you have successfully run `npm install`, in order to serve up the _client side_ of your project, you will run 
+`ng serve` (from the `client` directory as well). This will trigger the various tools in the
 client side portion of the project to build and host your client side
-application on their own little web-server, available by default at ``localhost:4200``. If your server is running, you will be able to see data for users if you navigate to the right place in the project.
+application on their own little web-server, available by default at [`localhost:4200`](http://localhost:4200). If your server is running, you will be able to see data for users if you navigate to the right place in the project.
 
 ## Testing and Continuous Integration
 
 There are now more testing options! You can test the client, or the server or both.
 
-Testing client:
+### Testing the client
+
+From the `client` directory:
+* `ng test` runs the client tests.
+* `ng test --code-coverage` runs the client tests and generates a coverage report
+  * It generates a coverage report you can find in your client directory `client/coverage/client/index.html`.
+  * Right click on `index.html` and select `Copy path` and paste it into your browser of choice. You can also drag and drop `index.html` onto the tab area of your browser and it will open it.
+
+### Testing the server
+
+From the `server` directory:
+* `./gradlew test` runs the server tests once.
+  * It generates a report you can find in `server/build/reports/tests/test/index.html`.
+* `./gradlew test jacocoTestReport` runs the server tests once and creates a coverage report
+  * It generates a coverage report you can find in `server/build/jacocoHtml/index.html` in addition to the regular report generated by the `test` task.
+
+### End to end testing with Cypress
+
+End to end (E2E) testing involves the whole software stack rather than one part of it. Our E2E tests look at the behavior of both the client and server and how they interact by simulating what a real user would do with it.
+
+We use [Cypress](https://www.cypress.io/) for our end-to-end tests. There are a few ways to run the E2E tests. They area all started from the `client` directory and require the server be running at the same time (`./gradlew run` in the `server` directory).
+
+* `ng e2e` both builds and serves the client and runs through all the Cypress end-to-end tests once.
+* `ng e2e --watch` builds and serves the client but just opens Cypress for you to be able to run the tests you want without closing automatically.
+  * This is the same as running `ng serve` and `npm run cy:open` (or `npx cypress open`) at the same time. If you are already running `ng serve` it will be easier to do this rather than closing it and running `ng e2e`.
```

Comment:
And they can't be running `ng server` twice (directly or indirectly) because multiple processing will be fighting for the port, right?

Reply:
Indeed, unlike when running `ng test`, the dev server started by `ng e2e` is the same as if started with `ng serve`. Attempting to run two instances at the same time will result in this:
![image](https://user-images.githubusercontent.com/1300395/108018640-eaa3f000-6fdd-11eb-8c6c-3609387a8e2c.png)